spec/models/user_spec.rb

```diff
@@ -0,0 +1,107 @@
+require "rails_helper"
+
+RSpec.describe User, type: :model do
+  let(:user) { FactoryBot.create :user }
+
+  describe "validates" do
+    it do
+      is_expected.to validate_presence_of(:name)
+      is_expected.to validate_presence_of(:email)
+      is_expected.to validate_presence_of(:password)
+      is_expected.to validate_length_of(:password).is_at_least(6)
+      is_expected.to define_enum_for(:role).with_values(%i(user admin))
+      is_expected.to define_enum_for(:status).with_values(%i(enabled disabled))
+    end
+
+    it do
+      user.name = "a" * 51
+      expect(user).not_to be_valid
+      expect(user.errors[:name]).to include("is too long (maximum is 50 characters)")
+    end
+
+    it do
+      user.name = nil
+      expect(user).not_to be_valid
+      expect(user.errors.size).to eq 1
+      expect(user.errors[:name]).to include("can't be blank")
+    end
+
+    it do
+      FactoryBot.create :user, email: "abc@gmail.com" 
+      user = FactoryBot.build :user, email: "abc@gmail.com" 
+      user.valid?
+      expect(user.errors.size).to eq 1
+      expect(user.errors[:email]).to include("has already been taken")
+    end
+
+    it do
+      FactoryBot.create :user, email: "abc@gmail.com" 
+      user = FactoryBot.build :user, email: "abc@gmail.com" 
+      user.valid?
+      expect(user.errors.size).to eq 1
+      expect(user.errors[:email]).to include("has already been taken")
+    end
+
+    it { expect(user.email.match(Settings.devise.email)).not_to be_nil }
+
+    it do
+      user.email = "a"
+      expect(user.email.match(Settings.devise.email)).to be_nil
+    end
+
+    it do
+      user = User.new
+      expect(user.role).to eq("user")
+    end
+
+    it do
+      user = User.new
+      expect(user.status).to eq("enabled")
+    end
+
+  end
+
+  describe "associations" do
+    it do
+      is_expected.to have_many(:orders)
+      is_expected.to have_many(:order_items).through(:orders)
+    end
+
+    it do
+      order = FactoryBot.create :order, user: user
+      order_item1 = FactoryBot.create :order_item, order: order
+      order_item2 = FactoryBot.create :order_item, order: order
+
+      expect(user.order_items).to include(order_item1, order_item2)
+    end
+
+    it do
+      order1 = FactoryBot.create :order, user: user
+      order2 = FactoryBot.create :order, user: user
+
+      expect(user.orders).to include(order1, order2)
+    end
+  end
+
+  describe "#confirmation_required?" do
+    it do
+      expect(user.confirmation_required?).to be true
+    end
+
+    it do
+      user.admin!
+      expect(user.confirmation_required?).to be false
+    end
+  end
```

Comment:
Thêm subject, nghiên cứu thêm về let và let!
```suggestion
  describe "#confirmation_required?" do
    subject { user.confirmation_required? }
    
   let(:user) { create :user, role: role }
   
   context "when user is admin" do
     let(:role) { "user" }
     
     # expectation
   end
  end
```